File: specs/Qowaiv.CodeAnalysis.VisualBasic.Specs/Cases/UseTestableTimeProvider.vb
﻿Imports System

Public Class DateTimeAsProvider

    Public Sub Noncompliant()
        Dim now As Date = DateTime.Now ' Noncompliant {{Use a testable (date) time provider instead.}}
        '                 ^^^^^^^^^^^^
        Dim utc As Date = DateTime.UtcNow ' Noncompliant
        Dim today As Date = DateTime.Today ' Noncompliant
        Dim _date As Date = Date.Now ' Noncompliant
    End Sub

    Public Sub CompliantAre()
        Dim other As Integer = DateTime.DaysInMonth(2000, 2) ' Compliant
        Dim noSystemDateTime As NotSystem.DateTime = NotSystem.DateTime.Now ' Compliant
        Dim noSystemtDate As NotSystem.Date = NotSystem.Date.Now ' Compliant
    End Sub
End Class

Namespace NotSystem
    Public Class DateTime
        Public Shared ReadOnly Property Now As DateTime
            Get
                Return New DateTime()
            End Get
        End Property
    End Class
    Public Class [Date]
        Public Shared ReadOnly Property Now As NotSystem.Date
            Get
                Return New NotSystem.Date()
            End Get
        End Property
    End Class
End Namespace

File: Specs/Cases/UseTestableTimeProvider.cs
using System;

public class DateTimeAsProvider
{
    public void Noncompliant()
    {
        var now = DateTime.Now; // Noncompliant {{Use a testable (date) time provider instead.}}
        //        ^^^^^^^^^^^^
        var utc = DateTime.UtcNow; // Noncompliant
        var today = DateTime.Today; // Noncompliant
        var offset_now = DateTimeOffset.Now; // Noncompliant
        var offset_utc = DateTimeOffset.UtcNow; // Noncompliant
    }

    public void CompliantAre()
    {
        var other = DateTime.DaysInMonth(2000, 2); // Compliant
        var noSystemDateTime = NotSystem.DateTime.Now; // Compliant
    }
}

namespace NotSystem
{
    public class DateTime
    {
        public static DateTime Now => new DateTime();
    }
}
